### ISSUE TITLE
Config in yaml/json format?

### ISSUE THREAD

**Issue by strkkk** [checkstyle/checkstyle#7246]
What do you think about adding additional config formats, e.g
Yaml
```
modules:
  NeedBracesCheck:
     id : testCheck
     severity : ERROR
```
Json
```
{
  "modules": [
    "NeedBracesCheck": {
       "id" : "testCheck",
       "severity" : "ERROR"
     }
  ]
}
```
?

**Comment by romani** [comment]
It will be hard for us to support two/few configs:
- we would need to use provide in documentation examples for all formats.
- we would need to have additional config (copy of xml) but in new format ("eat food that you produce").

We have problems even now to make sure that example in xdoc is actually working, there were bunch of issues on this, and there are still problems left.

ok, even we will find a way on how to update all our examples and ...

There are nuances in proposed formats:

JSON:
- json can have schema (https://json-schema.org/) - that is good.
- json do not have comments !!! (please look at out config, it is full of comments). yes, we can have special field "config-comment", but not sure it is good.
- there is no support for multiline string values (so there are workarounds https://stackoverflow.com/questions/2392766/multiline-strings-in-json) 
- support of duplicated fields ?  (in XML we do support multiple tag of the same name, that all values will be concatenated.)

YAML:
- there are comments - good
- there is no Schema (so users desperately search in web for examples) ... or I am not aware of it existence
- any extra space  and yaml is weird, hard to parse or parsed in way user do not expect
- there are multi-line support, but it is so error prone as users do not know nuances well - https://yaml-multiline.info/ (personal experience if also full of pain)

====

[JSON](http://www.json.org/) and [YAML](https://yaml.org/) are "lightweight data-interchange format" and "data serialization standard", NOT languages (XML is language). So that is why there are lot of nuances and differences. Language is by definition is more flexible and more powerful, and we already use some features, so to let users benefit we need to give them good substitution. 

even there is fashion to do all in such formats, people usually forget about nuances of day to day usage by users. Each format have its own prons and cons.

====

I can not find now ... a while ago I had conversation with users about groovy config support (based on fashion of converting configs from XML to Groovy (language to language conversion)). It is end up in something undecided as it is fun to implement new format parser, BUT it is not a fun to update documentation to let users see examples in both formats. See example of how it works for logback - http://logback.qos.ch/manual/appenders.html#ConsoleAppender.

**Comment by strkkk** [comment]
@romani I agree that comments are big disadvantage of json format, I think it can be excluded and only yaml is discussed.

About your concerns in general - user always have a way to choose which format is more acceptable to them. Support of the docs - yeah, this is a tough one. May be there is some automated tools that can help with it.

1. When schema can be useful?
2. About multilines - looks like an issue, but is there many multiline stuff in config? Only xpath and regex things come to my mind.

Groovy option from link you provided looks nice as well.

**Comment by romani** [comment]
> May be there is some automated tools that can help with it.

In our case we need so custom parsing of xml, we would need to have our own tool.

>When schema can be useful?

To show user where to put part of config. It is well known issue with yaml, is that you can explain in documentation some new field. But will be big question for user on what level to put it. The way to resolve this is only bunch of examples in documentation. Examples: ansimble, docker compose, all our CI config ( it is real struggle to make and modify config, but I agree it is easy to read it later). If you want to experience yaml, you can try to add azure CI in our project, there is issue on it.

>looks like an issue, but is there many multiline stuff in config?
Messages:
https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L148
Tokens: https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L183
Property: https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L306
In yaml, there is support for multiline, it is possible to do same, but it is painful process if you don't know all nuances.

The main problem with yaml is that it doesn't have schema, so all mistakes will be considered as new properties or ... . I have never seen good error about mistake in config, that cause me hours of training to make it work.
Yaml is good when some tool write it for you.
------

Even you figure out all nuances, and provide tool for xml to yaml conversation and write your own schema like validation ... Good ... But what about all plugins ? As most users use same config in build tool, ide, services, ... . So it will be extra fight to actually let user use it. And even this is doable.
But I will not participate in this, as we have problems that will let project die if not resolved in next year. Example: new syntax support in Java especially multiline strings. There are morebig tasks that will make support/contribution of project much easier, I would rather spend time on them.

If you see a way how to make new format introduction smooth, keep sharing thoughts

**Comment by romani** [comment]
@strkkk , should we close this discussion ?

**Comment by strkkk** [comment]
@romani ok

**Comment by romani** [comment]
As someone has good plan on how to make new format of config (details, limitations, ... ) - please create new issue.
